File: XZKit/Sources/Category/UIKit/UIColor+XZKit.m
//
//  UIColor.m
//  XZKit
//
//  Created by Xezun on 2017/10/24.
//

#define XZ_RGBA_COLOR
#import "UIColor+XZKit.h"

@implementation UIColor (XZKit)

- (XZColor)XZColor {
    CGFloat r = 0, g = 0, b = 0, a = 0;
    if ([self getRed:&r green:&g blue:&b alpha:&a]) {
        return XZColorMake(round(r * 255), round(g * 255), round(b * 255), round(a * 255));
    }
    return XZColorMake(0, 0, 0, 0);
}

+ (UIColor *)xz_colorWithXZColor:(XZColor)color {
    CGFloat const r = color.red   / 255.0;
    CGFloat const g = color.green / 255.0;
    CGFloat const b = color.blue  / 255.0;
    CGFloat const a = color.alpha / 255.0;
    return [UIColor colorWithRed:r green:g blue:b alpha:a];
}

+ (UIColor *)xz_colorWithRGB:(NSInteger)rgb {
    CGFloat const red   = (rgb>>16&0xFF) / 255.0;
    CGFloat const green = (rgb>> 8&0xFF) / 255.0;
    CGFloat const blue  = (rgb>> 0&0xFF) / 255.0;
    return [UIColor colorWithRed:red green:green blue:blue alpha:1.0];
}

+ (UIColor *)xz_colorWithRGBA:(NSInteger)rgba {
    CGFloat const red   = (rgba>>24&0xFF) / 255.0;
    CGFloat const green = (rgba>>16&0xFF) / 255.0;
    CGFloat const blue  = (rgba>> 8&0xFF) / 255.0;
    CGFloat const alpha = (rgba>> 0&0xFF) / 255.0;
    return [UIColor colorWithRed:red green:green blue:blue alpha:alpha];
}

+ (UIColor *)xz_colorWithRed:(NSInteger)red Green:(NSInteger)green Blue:(NSInteger)blue Alpha:(NSInteger)alpha {
    return [UIColor colorWithRed:red/255.0 green:green/255.0 blue:blue/255.0 alpha:alpha/255.0];
}

+ (instancetype)xz_colorWithString:(NSString *)string {
    XZColor const v = XZColorFromString(string);
    CGFloat const r = v.red   / 255.0;
    CGFloat const g = v.green / 255.0;
    CGFloat const b = v.blue  / 255.0;
    CGFloat const a = v.alpha / 255.0;
    return [UIColor colorWithRed:r green:g blue:b alpha:a];
}

+ (UIColor *)xz_colorWithString:(NSString *)string alpha:(CGFloat)alpha {
    XZColor const v = XZColorFromString(string);
    CGFloat const r = v.red   / 255.0;
    CGFloat const g = v.green / 255.0;
    CGFloat const b = v.blue  / 255.0;
    return [UIColor colorWithRed:r green:g blue:b alpha:alpha];
}

@end


XZColor XZColorFromString(NSString *string) {
    if (string.length < 3) {
        return XZColorMake(0, 0, 0, 0);
    }
    
    const char * const characters = [string cStringUsingEncoding:(NSASCIIStringEncoding)];
    char numbers[9] = {0};  // 存储数字。
    NSInteger count = 0;    // numbers 的长度。
    
    // 满 8 位就不用继续查找了
    for (NSInteger i = 0; count < 8; i++) {
        char const character = characters[i];
        
        // 到了末尾
        if (character == '\0') {
            break;
        }
        
        // A - F
        if (character >= 'A' && character <= 'F') {
            numbers[count++] = character - 'A' + 10;
            continue;
        }
        // a - f
        if (character >= 'a' && character <= 'f') {
            numbers[count++] = character - 'a' + 10;
            continue;
        }
        // 0 - 9
        if (character >= '0' && character <= '9') {
            numbers[count++] = character - '0';
            continue;
        }
        
        // 已解析到颜色值
        if (count >= 3) {
            break;
        }
        
        // 连续的数字不足 3 位，重新开始查找
        count = 0;
    }
    
    switch (count) {
        case 3: { // #ABC => #AABBCCFF
            CGFloat const r = numbers[0] * 16 + numbers[0];
            CGFloat const g = numbers[1] * 16 + numbers[1];
            CGFloat const b = numbers[2] * 16 + numbers[2];
            return XZColorMake(r, g, b, 255);
        }
        case 4: { // #ABCD => #AABBCCDD
            CGFloat const r = numbers[0] * 16 + numbers[0];
            CGFloat const g = numbers[1] * 16 + numbers[1];
            CGFloat const b = numbers[2] * 16 + numbers[2];
            CGFloat const a = numbers[3] * 16 + numbers[3];
            return XZColorMake(r, g, b, a);
        }
        case 5: { // #ABCDE => #AABBCCDE
            CGFloat const r = numbers[0] * 16 + numbers[0];
            CGFloat const g = numbers[1] * 16 + numbers[1];
            CGFloat const b = numbers[2] * 16 + numbers[2];
            CGFloat const a = numbers[3] * 16 + numbers[4];
            return XZColorMake(r, g, b, a);
        }
        case 6:{ // #123456 => #123456FF
            CGFloat const r = numbers[0] * 16 + numbers[1];
            CGFloat const g = numbers[2] * 16 + numbers[3];
            CGFloat const b = numbers[4] * 16 + numbers[5];
            return XZColorMake(r, g, b, 255);
        }
        case 7: { // #123456A => #123456AA
            CGFloat const r = numbers[0] * 16 + numbers[1];
            CGFloat const g = numbers[2] * 16 + numbers[3];
            CGFloat const b = numbers[4] * 16 + numbers[5];
            CGFloat const a = numbers[6] * 16 + numbers[6];
            return XZColorMake(r, g, b, a);
        }
        case 8: { // #123456AA => #123456AA
            CGFloat const r = numbers[0] * 16 + numbers[1];
            CGFloat const g = numbers[2] * 16 + numbers[3];
            CGFloat const b = numbers[4] * 16 + numbers[5];
            CGFloat const a = numbers[6] * 16 + numbers[7];
            return XZColorMake(r, g, b, a);
        }
        default: {
            return XZColorMake(0, 0, 0, 0);
        }
    }
}

File: XZKit/Sources/Category/UIKit/UIColor+XZKit.swift
//
//  UIColor+XZKit.swift
//  XZKit
//
//  Created by Xezun on 2021/2/14.
//

import Foundation

extension XZColor: ExpressibleByIntegerLiteral, RawRepresentable {
    
    public typealias IntegerLiteralType = Int
    
    public init(integerLiteral value: Int) {
        self.init(rawValue: value)
    }
    
    public typealias RawValue = Int
    
    public var rawValue: Int {
        return alpha + (blue << 8) + (green << 16) + (red << 24);
    }
    
    public init(rawValue: Int) {
        let r = (rawValue >> 24) & 0xFF;
        let g = (rawValue >> 16) & 0xFF;
        let b = (rawValue >>  8) & 0xFF;
        let a = (rawValue >>  0) & 0xFF;
        self.init(red: r, green: g, blue: b, alpha: a);
    }
    
    public var uiColor: UIColor {
        return UIColor(self)
    }
    
}


extension Int {
    
    public init(_ color: XZColor) {
        self = color.rawValue
    }
    
}

/// 通过字符串形式 RGBA 值构造 UIColor 对象。
/// - Parameter string: 十六进制颜色字符串，如 #A1B2C3FF
/// - Returns: UIColr
public func rgba(_ string: String) -> UIColor {
    let color = XZColor.init(string)
    return rgba(color.red, color.green, color.blue, color.alpha)
}

/// 通过整数形式的 RGBA 值构造 UIColor 对象。
/// - Parameter value: 十六进制 RGBA 颜色值，如 0xFF0000FF
/// - Returns: UIColor
public func rgba(_ value: Int) -> UIColor {
    let r = CGFloat((value >> 24) & 0xFF) / 255.0
    let g = CGFloat((value >> 16) & 0xFF) / 255.0
    let b = CGFloat((value >>  8) & 0xFF) / 255.0
    let a = CGFloat((value >>  0) & 0xFF) / 255.0
    return UIColor(red: r, green: g, blue: b, alpha: a)
}

/// 通过 RGBA 值 [0, 255] 构造 UIColor 对象。
public func rgba(_ red: Int, _ green: Int, _ blue: Int, _ alpha: Int) -> UIColor {
    let r = CGFloat(red)   / 255.0
    let g = CGFloat(green) / 255.0
    let b = CGFloat(blue)  / 255.0
    let a = CGFloat(alpha) / 255.0
    return UIColor(red: r, green: g, blue: b, alpha: a)
}

/// 通过 RGBA 值 [0, 1.0] 构造 UIColor 对象。
public func rgba(_ red: CGFloat, _ green: CGFloat, _ blue: CGFloat, _ alpha: CGFloat) -> UIColor {
    return UIColor(red: red, green: green, blue: blue, alpha: alpha)
}

/// 通过字符串形式 RGB 值构造 UIColor 对象。
/// - Parameter string: 十六进制颜色字符串，如 #A1B2C3FF
/// - Returns: UIColr
public func rgb(_ string: String) -> UIColor {
    let color = XZColor.init(string)
    return rgb(color.red, color.green, color.blue)
}

/// 通过整数形式的 RGB 值构造 UIColor 对象。
/// - Parameter value: 十六进制 RGBA 颜色值，如 0xFF0000
/// - Returns: UIColor
public func rgb(_ value: Int) -> UIColor {
    let r = value >> 16
    let g = value >> 8
    let b = value
    return rgba(r&0xFF, g&0xFF, b&0xFF, 255)
}

/// 通过 RGB 值 [0, 255] 构造 UIColor 对象。
public func rgb(_ red: Int, _ green: Int, _ blue: Int) -> UIColor {
    return rgba(red, green, blue, 255)
}

/// 通过 RGB 值 [0, 1.0] 构造 UIColor 对象。
public func rgb(_ red: CGFloat, _ green: CGFloat, _ blue: CGFloat) -> UIColor {
    return rgba(red, green, blue, 255)
}


